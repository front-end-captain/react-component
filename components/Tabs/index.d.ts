import Tabs from "./tabs";

export { TabsProps } from "./tabs";
export { TabPaneProps } from "./tabPane";

export default Tabs;
