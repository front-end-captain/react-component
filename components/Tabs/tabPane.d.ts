import * as React from "react";

export interface TabPaneProps {
  key: string;
  tab: string | React.ReactNode;
  disabled: boolean;
}

declare const TabPane: (props: TabPaneProps) => React.ReactNode;

export default TabPane;
