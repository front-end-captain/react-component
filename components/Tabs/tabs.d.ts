import * as React from "react";

import TabPane from "./tabPane";

interface TabsProps {
  prefixCls?: string;
  className?: string;
  style?: React.CSSProperties;
  tabBarStyle?: React.CSSProperties;

  type: "line" | "card";

  activeKey?: string;
  defaultActiveKey?: string;
  // animated?: boolean;

  // renderTabBar?: (props: TabsProps, defaultTabBar: React.ReactNode) => React.ReactNode;
  tabBarExtraContent?: React.ReactNode;

  onChange?: (activeKey: string) => void;
  // onNextClick?: (nextActiveKey: string) => void;
  // onPrevClick?: (nextActiveKey: string) => void;
  onTabClick?: (key: string) => void;
}

interface TabsState {
  activeKey: number,
}

export default class Tabs extends React.Component<TabsProps, TabsState> {
  static TabPane: typeof TabPane;
}
