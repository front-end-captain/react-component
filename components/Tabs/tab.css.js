/* stylelint-disable declaration-colon-newline-after */
/* stylelint-disable indentation */
import styled from "styled-components";

const TabsWrapper = styled.div`
  display: block;
`;

const TabsHeaderWrapper = styled.div`
  display: block;
  position: relative;
  outline: none;
  transition: padding 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);

  .${(props) => props.prefixCls}-nav {
    display: inline-block;
    width: auto;
    position: relative;

    .${(props) => props.prefixCls}-header-item {
      display: inline-block;
      box-sizing: border-box;
      height: 100%;
      margin: 0 32px 0 0;
      padding: 12px 16px;
      text-decoration: none;
      cursor: pointer;
      transition: color 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
    }

    .${(props) => props.prefixCls}-header-item-active {
      color: #1890ff;
      font-weight: 500;
    }

    .${(props) => props.prefixCls}-header-item-disabled {
      cursor: not-allowed;
      color: rgba(0, 0, 0, 0.25);
    }

    .${(props) => props.prefixCls}-ink-line {
      position: absolute;
      bottom: -1px;
      left: 0;
      z-index: 1;
      box-sizing: border-box;
      height: 2px;
      background-color: #1890ff;
      transform-origin: 0 0;
      transition: transform 0.3s cubic-bezier(0.645, 0.045, 0.355, 1),
        width 0.3s cubic-bezier(0.645, 0.045, 0.355, 1),
        left 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
    }

    &.${(props) => props.prefixCls}-nav-card {
      .${(props) => props.prefixCls}-header-item {
        margin: 0;
        margin-right: 2px;
        padding: 0 16px;
        line-height: 38px;
        border: 1px solid #e8e8e8;
        border-bottom: 0;
        border-radius: 4px 4px 0 0;
      }

      .${(props) => props.prefixCls}-ink-line {
        visibility: hidden;
      }

      .${(props) => props.prefixCls}-header-item-active {
        margin-bottom: -1px;
        color: #1890ff;
        background: #fff;
        border-color: #e8e8e8;
        border-bottom: 1px solid transparent;
      }
    }
  }
`;

const TabsPaneWrapper = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
  transition: margin-left 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
  will-change: margin-left;

  .${(props) => props.prefixCls}-pane-item {
    display: inline-block;
    flex-shrink: 0;
    width: 100%;
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid #e8e8e8;
  }

  .${(props) => props.prefixCls}-pane-item-active {
    opacity: 1;
    transition: opacity 0.3s;
  }

  .${(props) => props.prefixCls}-pane-item-inactive {
    height: 0;
    padding: 0 !important;
    opacity: 0;
    pointer-events: none;
  }
`;

export { TabsWrapper, TabsHeaderWrapper, TabsPaneWrapper };
