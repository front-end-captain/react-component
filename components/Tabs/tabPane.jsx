import React from "react";
import classNames from "classnames";

const TabPane = ({ children, isActive, prefixCls, disabled }) => {
  const paneClassName = classNames(`${prefixCls}-pane-item`, {
    [`${prefixCls}-pane-item-active`]: isActive,
    [`${prefixCls}-pane-item-inactive`]: !isActive,
    [`${prefixCls}-pane-item-disabled`]: disabled,
  });

  return <div className={paneClassName}>{children}</div>;
};

export default TabPane;
