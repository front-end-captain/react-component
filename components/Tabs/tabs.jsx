import React, { Fragment, Children, cloneElement } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";
import { isFragment } from "react-is";

import TabPane from "./tabPane.jsx";

import { TabsWrapper, TabsPaneWrapper, TabsHeaderWrapper } from "./tab.css.js";

class Tabs extends React.Component {
  static defaultProps = {
    prefixCls: "sky-tab",
    defaultActiveKey: 1,
    type: "line",
    onChange: () => {},
    onTabClick: () => {},
  };

  static propTypes = {
    prefixCls: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.objectOf(PropTypes.string),
    type: PropTypes.oneOf(["line", "card"]),
    tabBarStyle: PropTypes.objectOf(PropTypes.string),
    activeKey: PropTypes.string,
    defaultActiveKey: PropTypes.string,
    tabBarExtraContent: PropTypes.element,
    onChange: PropTypes.func,
    onTabClick: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.tabsHeader = React.createRef();

    this.state = {
      activeKey: Number(props.activeKey) || Number(props.defaultActiveKey),
      inkLineWidth: 0,
      inkLineOffsetLeft: 0,
    };
  }

  componentDidMount() {
    this.setInkLinePosition();
  }

  handleChild = (child, index) => {
    if (!child) {
      return null;
    }

    const { activeKey } = this.state;
    const key = child.key || String(index);
    const { tab, children, disabled } = child.props;
    const { prefixCls } = this.props;

    const isActive = activeKey === Number(key);

    const props = {
      key,
      tab,
      isActive,
      prefixCls,
      disabled,
      children,
    };

    return cloneElement(<TabPane>{child}</TabPane>, props);
  };

  getItem = () => {
    const { children } = this.props;
    const newChildren = Children.map(this.getChildren(), this.handleChild);

    if (isFragment(children)) {
      return <Fragment>{newChildren}</Fragment>;
    }

    return newChildren;
  };

  getHeader = () => {
    const { activeKey } = this.state;
    const headers = Children.map(this.getChildren(), (child, index) => {
      const key = child.key || String(index);
      const { tab, disabled } = child.props;
      const { prefixCls } = this.props;

      const isActive = activeKey === Number(key);

      const bindRef = isActive ? { ref: (node) => (this.activeTab = node) } : {};

      const props = {
        className: classNames(`${prefixCls}-header-item`, {
          [`${prefixCls}-header-item-active`]: isActive,
          [`${prefixCls}-header-item-disabled`]: disabled,
        }),
        key,
        onClick: disabled ? undefined : () => this.setActiveKey(key),
      };

      return cloneElement(<div {...bindRef}>{tab}</div>, props);
    });

    return headers;
  };

  getChildren = () => {
    const { children } = this.props;
    const realChildren = isFragment(children) ? children.props.children : children;
    return realChildren;
  };

  setActiveKey = (key) => {
    const { onChange, onTabClick } = this.props;
    const { activeKey } = this.state;

    onTabClick();

    if (activeKey === Number(key)) {
      return;
    }

    this.setState({ activeKey: Number(key) }, () => {
      this.setInkLinePosition();
      onChange(key);
    });
  };

  setInkLinePosition = () => {
    const { width, left } = (this.activeTab && this.activeTab.getBoundingClientRect()) || {};
    const { left: headerOffset } =
      (this.tabsHeader.current && this.tabsHeader.current.getBoundingClientRect()) || {};

    this.setState({
      inkLineWidth: width,
      inkLineOffsetLeft: left - headerOffset,
    });
  };

  render() {
    const { activeKey, inkLineWidth, inkLineOffsetLeft } = this.state;
    const { className, style, prefixCls, type, tabBarStyle, tabBarExtraContent } = this.props;

    const wrapperClassName = classNames(`${prefixCls}-wrapper`, className);
    const navClassName = classNames(`${prefixCls}-nav`, `${prefixCls}-nav-${type}`);

    return (
      <TabsWrapper style={style} prefixCls={prefixCls} className={wrapperClassName}>
        <TabsHeaderWrapper prefixCls={prefixCls} ref={this.tabsHeader} style={tabBarStyle}>
          <div className={navClassName}>
            {this.getHeader()}
            <div
              className={`${prefixCls}-ink-line`}
              style={{
                width: inkLineWidth,
                transform: `translate3d(${inkLineOffsetLeft}px, 0, 0)`,
              }}
            />
          </div>
          {tabBarExtraContent && <div style={{ float: "right" }}>{tabBarExtraContent}</div>}
        </TabsHeaderWrapper>
        <TabsPaneWrapper prefixCls={prefixCls} style={{ marginLeft: `-${(activeKey - 1) * 100}%` }}>
          {this.getItem()}
        </TabsPaneWrapper>
      </TabsWrapper>
    );
  }
}

Tabs.TabPane = TabPane;

export default Tabs;
