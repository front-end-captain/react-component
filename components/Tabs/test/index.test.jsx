import React from "react";
import { render } from "enzyme";
import toJson from "enzyme-to-json";

import Tabs from "./../index.js";

const { TabPane } = Tabs;

describe("<Tab />", () => {
  it("should render some Tabs components", () => {
    const wrapper = render(
      <div>
        <Tabs
          defaultActiveKey="1"
          onChange={(key) => console.log(key)}
          onTabClick={() => console.log("tab clicked")}
        >
          <TabPane tab="Tab 1" key="1">
            Content of Tab Pane 1
          </TabPane>
          <TabPane tab="Tab 2" key="2">
            Content of Tab Pane 2
          </TabPane>
          <TabPane tab="Tab 3" key="3">
            Content of Tab Pane 3
          </TabPane>
        </Tabs>
        <Tabs defaultActiveKey="2" type="card">
          <TabPane tab="Tab 1" key="1">
            Content of Tab Pane 1
          </TabPane>
          <TabPane tab="Tab 2" key="2">
            Content of Tab Pane 2
          </TabPane>
          <TabPane tab="Tab 3" key="3">
            Content of Tab Pane 3
          </TabPane>
        </Tabs>
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
