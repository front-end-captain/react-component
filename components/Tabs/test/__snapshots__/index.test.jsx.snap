// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> should render some Tabs components 1`] = `
<div>
  <div
    class="sky-tab-wrapper tabcss__TabsWrapper-t8bu7f-0 dNCAjj"
  >
    <div
      class="tabcss__TabsHeaderWrapper-t8bu7f-1 hPfwkj"
    >
      <div
        class="sky-tab-nav sky-tab-nav-line"
      >
        <div
          class="sky-tab-header-item sky-tab-header-item-active"
        >
          Tab 1
        </div>
        <div
          class="sky-tab-header-item"
        >
          Tab 2
        </div>
        <div
          class="sky-tab-header-item"
        >
          Tab 3
        </div>
        <div
          class="sky-tab-ink-line"
          style="width:0;transform:translate3d(0px, 0, 0)"
        />
      </div>
    </div>
    <div
      class="tabcss__TabsPaneWrapper-t8bu7f-2 hCcAuF"
      style="margin-left:-0%"
    >
      <div
        class="sky-tab-pane-item sky-tab-pane-item-active"
      >
        Content of Tab Pane 1
      </div>
      <div
        class="sky-tab-pane-item sky-tab-pane-item-inactive"
      >
        Content of Tab Pane 2
      </div>
      <div
        class="sky-tab-pane-item sky-tab-pane-item-inactive"
      >
        Content of Tab Pane 3
      </div>
    </div>
  </div>
  <div
    class="sky-tab-wrapper tabcss__TabsWrapper-t8bu7f-0 dNCAjj"
  >
    <div
      class="tabcss__TabsHeaderWrapper-t8bu7f-1 hPfwkj"
    >
      <div
        class="sky-tab-nav sky-tab-nav-card"
      >
        <div
          class="sky-tab-header-item"
        >
          Tab 1
        </div>
        <div
          class="sky-tab-header-item sky-tab-header-item-active"
        >
          Tab 2
        </div>
        <div
          class="sky-tab-header-item"
        >
          Tab 3
        </div>
        <div
          class="sky-tab-ink-line"
          style="width:0;transform:translate3d(0px, 0, 0)"
        />
      </div>
    </div>
    <div
      class="tabcss__TabsPaneWrapper-t8bu7f-2 hCcAuF"
      style="margin-left:-100%"
    >
      <div
        class="sky-tab-pane-item sky-tab-pane-item-inactive"
      >
        Content of Tab Pane 1
      </div>
      <div
        class="sky-tab-pane-item sky-tab-pane-item-active"
      >
        Content of Tab Pane 2
      </div>
      <div
        class="sky-tab-pane-item sky-tab-pane-item-inactive"
      >
        Content of Tab Pane 3
      </div>
    </div>
  </div>
</div>
`;
